lsp-mode.el

```diff
@@ -4429,44 +4429,51 @@ Also, additional data to attached to each candidate can be passed via PLIST."
-                          'lsp-completion-start-point start-point
-                          'lsp-completion-prefix-line prefix-line)
-                  (text-properties-at 0 candidate))
-                 ((&hash "label"
-                         "insertText" insert-text
-                         "textEdit" text-edit
-                         "insertTextFormat" insert-text-format
-                         "additionalTextEdits" additional-text-edits)
-                  item))
-           (cond
-            (text-edit
-             (delete-region (point-at-bol) (point))
-             (insert prefix-line)
-             (lsp--apply-text-edit text-edit))
-            ((or insert-text label)
-             (delete-region (point-at-bol) (point))
-             (insert prefix-line)
-             (delete-region start-point (point))
-             (insert (or insert-text label))))
-
-           (when (eq insert-text-format 2)
-             (yas-expand-snippet
-              (lsp--to-yasnippet-snippet (buffer-substring start-point (point)))
-              start-point
-              (point)))
-           (when (and lsp-completion-enable-additional-text-edit additional-text-edits)
-             (lsp--apply-text-edits additional-text-edits)))
-         (lsp--capf-clear-cache)
-         (when (and lsp-signature-auto-activate
-                    (lsp-feature? "textDocument/signatureHelp"))
-           (lsp-signature-activate))
-
-         (setq-local lsp-inhibit-lsp-hooks nil)
-
-         (when (lsp--looking-back-trigger-characterp trigger-chars)
-           (setq this-command 'self-insert-command)))))))
+       (-rpartial #'lsp--capf-exit-fn trigger-chars)))))
+
+(defun lsp--capf-exit-fn (candidate _status &optional trigger-chars)
+  "Exit function of `completion-at-point'.
+CANDIDATE is the selected completion item.
+Others: TRIGGER-CHARS"
+  (-let* (((&plist 'lsp-completion-item item
+                   'lsp-completion-start-point start-point
+                   'lsp-completion-markers markers)
+           (text-properties-at 0 candidate))
+          ((&hash "label"
+                  "insertText" insert-text
+                  "textEdit" text-edit
+                  "insertTextFormat" insert-text-format
+                  "additionalTextEdits" additional-text-edits)
+           item))
+    (cond
+     (text-edit
+      (apply #'delete-region markers)
+      (lsp--apply-text-edit text-edit))
+     ((or insert-text label)
+      (apply #'delete-region markers)
+      (delete-region start-point (point))
+      (insert (or insert-text label))))
+    (-some-> markers (cadr) (set-marker nil))
+
+    (when (eq insert-text-format 2)
+      (yas-expand-snippet
+       (lsp--to-yasnippet-snippet (buffer-substring start-point (point)))
+       start-point
+       (point)))
+
+    (when (and lsp-completion-enable-additional-text-edit additional-text-edits)
+      (lsp--apply-text-edits additional-text-edits)))
+
+  (lsp--capf-clear-cache)
```

Comment:
seems like clear-cache is a better place to clean up the markers, isn't it?

Reply:
It's a little bit tricky since `post-completion` is being run after `company-completion-finished-hook`, thus required the marker is not being clean until after `post-completion`.
I've modified company hook to only clean cache in `company-after-completion-hook`.